Make plugin independent of Jooq version

oehme opened the pull request:
This change allows users to choose their desired jOOQ version and edition on the JooqExtension. It also replaces the in-process execution of jOOQ with a forked execution using a dedicated classpath.

This has many benefits:

- users can use whatever Jooq version they want, independent of which this plugin is compiled against
- users don't need to use `buildscript` workarounds to adjust the version, so they can instead use the much more concise `plugins` DSL
- the plugin integrates with `java-base` instead of `java` now, so it is easier to use for projects/enterprises with their own conventions

oehme commented:
@lukaseder I ended up pretty much finishing the whole story after all. One problem though: GenerationTool.main does not provide an exit code when it fails, so the build will just continue. Can you fix that on the Jooq side?

oehme commented:
This fixes #12 and supersedes #14 and #16

lukaseder commented:
@oehme. Hmm, I agree that we could improve that. There are some conditions under which the `GenerationTool` exits with error code `-1`, right now.

However, I'm thinking that perhaps it's not a good idea to exit with an error code at all, because the `GenerationTool` might be used by a client JVM that uses programmatic configuration. Would an exception work as well for you? What's your main entry point? `main(String[])` or `main(Configuration)`?

oehme commented:
We fork a new process, so our entry point is `main(String...)`. I don't
really care whether it uses an exit code or an exception, as long as it
indicates failure :) An exit code and possibly some usage instructions are
more user friendly than printing long stacktraces though.

Speaking of stack traces: I saw exceptions getting logged twice, so you
might want to check the whole error handling chain.

Programmatic clients will use `new GenerationTool().run(config)`, so they
wouldn't be affected by a change to the main method.

Am 04.09.2016 3:41 Nachm. schrieb "Lukas Eder" <notifications@github.com>:

@oehme <https://github.com/oehme>. Hmm, I agree that we could improve that.
There are some conditions under which the GenerationTool exits with error
code -1, right now.

However, I'm thinking that perhaps it's not a good idea to exit with an
error code at all, because the GenerationTool might be used by a client JVM
that uses programmatic configuration. Would an exception work as well for
you? What's your main entry point? main(String[]) or main(Configuration)?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/etiennestuder/gradle-jooq-plugin/pull/20#issuecomment-244621311>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABIvderMjGh2aXHyW9fIyP8T8Kp0IbZPks5qmxDzgaJpZM4Jzoi1>
.

aram535 commented:
Just a suggestions but if we're going to catch up the plugin ... why upgrade to gradle 2.5?  Plugins should be compatible with the latest version (3.0 as of this post).  Also gradle directory should be added to .gitignore at some point since it really shouldn't be checked in and updated per commit.

lukaseder commented:
@oehme OK, thanks for your feedback. Will throw exceptions from non-main methods and catch them / translate them to exit codes in `main(String[])`

oehme commented:
The plugin will be compatible with 2.5+.

The gradle directory must be checked in, it contains the wrapper properties
needed to start Gradle.

Am 05.09.2016 6:50 Nachm. schrieb "aram535" <notifications@github.com>:

> Just a suggestions but if we're going to catch up the plugin ... why
> upgrade to gradle 2.5? Plugins should be compatible with the latest version
> (3.0 as of this post). Also gradle directory should be added to .gitignore
> at some point since it really shouldn't be checked in and updated per
> commit.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/etiennestuder/gradle-jooq-plugin/pull/20#issuecomment-244814784>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABIvde662fslbWQ1WipQQFLABtMdRU8Cks5qnI6JgaJpZM4Jzoi1>
> .
>

aram535 commented:
@oehme I stand corrected.  I didn't realize you had increased the distributionUrl as well.  Can we use the -bin version of the URL since there is no need for the gradle source code and documentation?

oehme commented:
The sources are very useful for developing the plugin.

Am 06.09.2016 9:09 Vorm. schrieb "aram535" <notifications@github.com>:

> @oehme <https://github.com/oehme> I stand corrected. I didn't realize you
> had increased the distributionUrl as well. Can we use the -bin version of
> the URL since there is no need for the gradle source code and documentation?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/etiennestuder/gradle-jooq-plugin/pull/20#issuecomment-244930667>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABIvdb-pMTX2dvJhf2o-K3eCOHzbyI07ks5qnVfogaJpZM4Jzoi1>
> .
>

aram535 commented:
Good point, I keep thinking of the use of the plugin which has nothing to do with the gradle toolset in the plugin itself.  Ignore this thread. :-)

sargue commented:
Hi people,

Nice addition. Just one question because it's a problem I am facing right now.

Would this new version allow the use of commercial jOOQ artifacts? FYI, the group changes from `org.jooq` to `org.jooq.pro`.

oehme commented:
It would, by setting the jooqClasspath on the jooq task. But we could
probably make that more convenient.

Are there more use cases where you want something else than jooq.generator
on the classpath? Are there any jooq generator extensions for instance?

Am 08.09.2016 06:27 schrieb "Sergi Baila" <notifications@github.com>:

> Hi people,
>
> Nice addition. Just one question because it's a problem I am facing right
> now.
>
> Would this new version allow the use of commercial jOOQ artifacts? FYI,
> the group changes from org.jooq to org.jooq.pro.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/etiennestuder/gradle-jooq-plugin/pull/20#issuecomment-245543340>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABIvdZspG1UBpjV6lwqM7s98PQRvnFWpks5qn9UZgaJpZM4Jzoi1>
> .
>

sargue commented:
Er... extensions? Didn't know what those are, sorry. The only other dependecy would be the JDBC drivers but you have that covered.

oehme commented:
@lukaseder Could you provide some insight in what else users might want on the generator classpath? So far we have the jdbc drivers and the jooq.generator jar. @sargue pointed out there is a pro version of the generator. Anything else I'm missing?

lukaseder commented:
Thanks for pinging, @oehme. The different groupIds for commercial artefacts are documented here:
http://www.jooq.org/doc/latest/manual/getting-started/tutorials/jooq-in-7-steps/jooq-in-7-steps-step1

Essentially, we have:

- `org.jooq` for the Open Source Edition (available from Maven Central)
- `org.jooq.pro` for the Express, Professional, and Enterprise Editions (available only from us)
- `org.jooq.pro-java-6` for the Express, Professional, and Enterprise Editions with Java 6 support (available only from us)
- `org.jooq.trial` for the free trial

Let me know if you'd like to have a license key for your integration tests, and I will be very happy to provide you with one.

Hmm, what else. There's also the `jooq-meta-extensions` artifact (https://github.com/jOOQ/jOOQ/tree/master/jOOQ-meta-extensions), which allows for integrating the code generator with JPA

lukaseder commented:
I should add that these groupIds were added to jOOQ 3.8.0 and 3.7.1 only. jOOQ 3.6 and less only had a single groupId `org.jooq` for all different editions, in case that matters for you

oehme commented:
Thanks Lukas, I might rethink how the jooqClasspath works based on this. Currently it looks on the compile classpath to find the jooq jar and figures out the version based on that. Then it resolves that version of the generator.

But based on your feedback a more flexible solution seems to be needed. Probably let the user specify the jooq version explicitly and then force that version across both generator and compile classpath. Then make the jooq.generator jar the default, but make it easy to override with the jooq.pro.generator for instance.

sargue commented:
I have found a workaround for using jOOP commercial edition with the plugin tweaking the `classpath` configuration.

    configurations.classpath {
        resolutionStrategy {
            dependencySubstitution {
                substitute module("org.jooq:jooq:3.6.2") with module("org.jooq.pro:jooq:${versions.jooq}")
                substitute module("org.jooq:jooq-meta:3.6.2") with module("org.jooq.pro:jooq-meta:${versions.jooq}")
                substitute module("org.jooq:jooq-codegen:3.6.2") with module("org.jooq.pro:jooq-codegen:${versions.jooq}")
            }
        }
    }

The `${versions.jooq}` is a property I have defined at the top of the file.

The problem also is that the `org.jooq.pro` artifacts are not found on public repositories. You have two options here: if you have a private repository just install the dependencies there and add the repository to your `buildscript`.

I don't have a private repository so I've opted for the second solution. I have created a `libs` directory with the structure of a Maven repository. Actually, I have installed the artifacts on my local maven repository (`~\.m2\repository`) and then copied them from there. This is how it looks like:

![image](https://cloud.githubusercontent.com/assets/333808/18380363/38756832-7677-11e6-85c7-2fd83e8ffb53.png)

Then I just need to add it as a repository.

This is my `buildscript` section in its enterity:

```
buildscript {
    ext.versions = [
        jooq: '3.8.4',
        guice: '4.1.0',
        jersey: '2.23.2'
    ]

    repositories {
        maven {
            url uri('libs')
        }
        jcenter()
    }

    dependencies {
        classpath 'nu.studer:gradle-jooq-plugin:1.0.6'
        classpath 'org.mariadb.jdbc:mariadb-java-client:1.5.1-RC'
        classpath 'net.sourceforge.jtds:jtds:1.3.1'
    }

    configurations.classpath {
        resolutionStrategy {
            dependencySubstitution {
                substitute module("org.jooq:jooq:3.6.2") with module("org.jooq.pro:jooq:${versions.jooq}")
                substitute module("org.jooq:jooq-meta:3.6.2") with module("org.jooq.pro:jooq-meta:${versions.jooq}")
                substitute module("org.jooq:jooq-codegen:3.6.2") with module("org.jooq.pro:jooq-codegen:${versions.jooq}")
            }
        }
    }
}
```

aram535 commented:
gradle.properties is, I believe, designed to handle variable data such as this. 

Something simple lkike:
jooqPackage=org.jooq 
jooqVersion=3.8.4 
databaseJar=com.mysql.jdbc.Driver 

or maybe even make it a dependency via 
databaseDep=CreateJarDependency('mysql','mysql-connector-jara','6.0.3')

oehme commented:
@aram535 I guess you are referring to the ext properties? That's a matter of preference, no rule or best practice.

I personally use gradle.properties very little. It breaks the reading flow if you extract every little thing as a property. I wouldn't put stuff like the jdbc driver in there for instance. I only use it for stuff that regularly needs to change and which really only requires a property. With everything else I tend to use more structured objects. You'll see that the Gradle build itself for instance has no gradle.properties at all. 

But if you prefer properties, you can of course go for that. Whatever works best for you and your build :)

lukaseder commented:
Having absolutely no idea how Gradle really works, isn't there the possibility of injecting a dependency to a Gradle plugin, just like you can inject a dependency to a Maven plugin? As in: A full dependency specification?

Even better:

1. If a specific jOOQ dependency is specified for the plugin, use that.
2. If there is a jOOQ dependency (any groupId: OSS, commercial, etc.) on the classpath, take that.
3. Otherwise, pull the latest jOOQ dependency from Maven Central as the default behaviour.

oehme commented:
That's exactly what I proposed @lukaseder. 

You don't inject it into the plugin itself though. The plugin is independent of jooq, it forks jooq in a new process with its own, clean classpath.

But we can do even better. When the user says he wants the pro version, we can force usage of the pro artifacts everywhere, including transitive dependencies the user might have. Otherwise you might end up with a weird mix. So I'm inclined to explicitly model the "edition", maybe in another pull request.

aram535 commented:
@oehme we'll have to agree to disagree on that one.  I believe, having a property file like this is exactly for this scenario. The property file can be checked in, then ignored so that the user who clone the repo and builds can manipulate the environment to match their requirement without having to worry about sync issues when there are updates.

oehme commented:
@lukaseder another thing to note is that there is not "the classpath". A Gradle Java project can have arbitrary source sets, each with their own classpath. Many projects use the standard 'java' plugin which provides a 'main' and a 'test' source set. But that's just a convention, not hardcoded like in Maven.

So we might want an "opinionated" jooq plugin which applies the 'java'  plugin and puts jooq on the main classpath. That's what probably 90% of users want. Everyone else could use an "unopinionated" jooq-base plugin which leaves the dependency definition to the user.

oehme commented:
@aram535 Environment specific properties should go into the gradle.properties file in your Gradle user home. The gradle.properties in your project dir are meant to be shared and kept in sync. If you have further questions, don't hesitate to start a thread on the Gradle forum :)

sargue commented:
Meanwhile not a single word from @etiennestuder , the owner of this repository.

Should we port this to maybe jooq group. Or @oehme, you should fork it there and continue working on it. Both @etiennestuder  and you work at Gradle Inc.?

oehme commented:
@etiennestuder is aware of this PR, but as you have seen above there are some jooq subtleties to consider, so this is not yet ready to merge anyway.

Once it is ready, I'm sure Etienne will merge it quickly and release a new version.

aram535 commented:
@oehme Sorry but saying it with authority doesn't make it the right answer.  Local gradle files are for system parameters not for passing Dev variables.

lukaseder commented:
> That's exactly what I proposed @lukaseder.

OK, my bad - I guess I didn't get that. Thanks for the other clarifications

sargue commented:
If I may introduce another use case for this plugin...

Sometimes custom code needs to be compiled and added to the classpath of the plugin. I'm thinking about custom code generation, for example.

http://www.jooq.org/doc/3.8/manual/code-generation/codegen-generatorstrategy/

So I can have a `net.sargue.jooq.MyCustomNameGeneratorStrategy` java class which needs to be compiled and added to the plugin classpath. Maybe that isn't even possible at that point. Maybe it is. The gradle experts will know.

lukaseder commented:
For the record, what @sargue mentions is solved in Maven in a bit of a cumbersome way, i.e. by adding another Maven module/project and letting the module/project that uses the jOOQ code generator depend on the former.

Indeed, if it's possible to improve this with Gradle, I would imagine it would be very useful.

oehme commented:
That will be possible with the new approach. All you'll need to do is put your custom code in a separate sourceSet and add that to the jooqClasspath, roughly like this:

```
sourcSets {
  jooq {
    java.srcDir 'src/jooq/java'
  }
}

tasks.withType(JooqGenerate) {
  jooqClasspath += sourcSets.jooq.output
}
```

If this is a common scenario, we could even make that part of the plugin, so src/jooq/java works out of the box.

lukaseder commented:
> If this is a common scenario

It is!

>we could even make that part of the plugin, so src/jooq/java works out of the box.

Very interesting. Or would it be reasonable to call this directory `src/jooq-codegen/java` instead? As the relevant code will depend on the `jooq-codegen` artifact...

sargue commented:
+1 for the `src/jooq-codegen/java` option, sounds really nice

oehme commented:
@lukaseder @sargue Thank you for the great feedback. I have incorporated it into the latest revision. Could you review the readme and make sure it is clear to you how the plugin will work? I have left the custom GeneratorStrategy proposal out of this PR. We can do that in another iteration. It will be very easy once this is merged.

@etiennestuder I've finished what I think should be part of this PR. There have been other proposals like the custom generator extensions, but I think we should do those in a separate PR. Can you have a look so we can get this into user's hands? :)

For reference, this is what a typical project will look like now:

```
plugins {
  id 'java'
  id 'nu.studer.jooq' version '2.0.0'
}

dependencies {
  // generator dependencies are now added using the normal dependencies block
  jooqGeneratorClasspath 'org.postgresql:postgresql:9.4-1206-jdbc42'
  
  // jooq dependency versions are auto-managed by the plugin, you can omit the version here
  compile 'org.jooq:jooq'
}

jooq {
    version = '3.8.2' // default, can be omitted
    edition = 'OSS'  // default, can be omitted

    //everything below here is unchanged from previous versions
    demo(sourceSets.main) {
        jdbc {
            driver = 'org.postgresql.Driver'
            url = 'jdbc:postgresql://localhost:5432/jooq-demo'
            user = 'admin'
            password = ''
            schema = 'public'
        }
        generator {
            name = 'org.jooq.util.DefaultGenerator'
            strategy {
                name = 'org.jooq.util.DefaultGeneratorStrategy'
            }
            database {
                name = 'org.jooq.util.postgres.PostgresDatabase'
                inputSchema = 'public'
            }
            generate {
                relations = true
                deprecated = false
                records = true
                immutablePojos = false
                fluentSetters = false

            }
            target {
                packageName = 'my.jooqdemo.generated'
            }
        }
    }
}
```

sargue commented on README.md:
````diff
@@ -10,55 +10,64 @@ participate in the Gradle uptodate checks.
+```groovy
+jooq {
+  version = '3.8.2' //the default
+  edition = 'OSS'   //the default
````
We should list the exact strings used for all the versions as a reference. 'PRO' is probably easy to guess but just to be thorough.

oehme replied:
Agreed.

sargue commented:
Any progress on this?

lukaseder commented:
@oehme @etiennestuder Hi folks. What's the status on all this? I keep helping poor jOOQ users understand the limitations of this plugin, which has a hard reference to jOOQ 3.6.2 which is tough to override.

Let me set this clear. This is a *major* support incident / noise producer for me (roughly 40 incidents in the past 3 months!). I would really like to spend my time on other stuff than supporting this plugin and its current limitations. I appreciate that this all started with https://github.com/jOOQ/jOOQ/pull/5415 where there was a promise of less support work on the jOOQ side once this plugin is ready, but right now this problem is damaging the reputation of jOOQ as I can tell. Can we please set a binding deadline when this will work for jOOQ users?

If @etiennestuder has more pressing work to do than merging and publishing @oehme's changes, what would be an alternative path to success?

Please advise.

aram535 commented:
@lukaseder I suggest you fork this repo so that it's under your control.  Then you can "allow" external developer to help you maintain it without it taking away from your time.

lukaseder commented:
@aram535: That's one option of course, although I'm not that big of a fan of "unfriendly forks".

See, if anything here would be under my control, then I would just suggest calling the standalone code generator, which was suggested to be a bad idea in https://github.com/jOOQ/jOOQ/pull/5415 ;) But I'm personally not unhappy to revert to standalone code gen calls - but let's give this another chance. I think @oehme's work looks very promising. All that's left to do is merging and publishing, right?

aram535 commented:
@lukaseder Good point ... I figured at some point Etienne who I believe is a VP at Gradle, would have the time to merge your fork back in - but that's probably a long shot since he hasn't replied to any of the notes here.  Is anyone a write member that can do the merge?

lukaseder commented:
Well, if anything at all, I wouldn't mind a clear statement by @etiennestuder on whether he's willing to continue maintenance, or if he prefers to contribute this plugin to jOOQ... Or merge @oehme's suggestion, of course :)

etiennestuder commented:
@lukaseder I understand your concerns and apologize for the delay. I've been following the discussions all along. I will have reviewed and merged the changes of the PR by Oct 28th.

lukaseder commented:
Thank you very much, @etiennestuder, I really appreciate it. Again, if this is a burden on your side, I'm very willing to discuss alternatives.

On the flip side, given this comment here:

> This is a major support incident / noise producer for me (roughly 40 incidents in the past 3 months!).

This just also means that since we changed the manual according to @oehme's feedback, the plugin is quite widely adopted, which is great news! :)

etiennestuder commented:
Thanks for the PR, @oehme ;-)
